les/vflux/server/prioritypool.go

```diff
@@ -500,3 +519,147 @@ func (pp *PriorityPool) updatePriority(node *enode.Node) {
 	}
 	updates = pp.tryActivate()
 }
+
+// CapacityCurve is a snapshot of the priority pool contents in a format that can efficiently
+// estimate how much capacity could be granted to a given node at a given priority level.
+type CapacityCurve struct {
+	points       []curvePoint       // curve points sorted in descending order of priority
+	index        map[enode.ID][]int // curve point indexes belonging to each node
+	exclude      []int              // curve point indexes of excluded node
+	excludeFirst bool               // true if activeCount == maxCount
+}
+
+type curvePoint struct {
+	freeCap uint64 // available capacity and node count at the current priority level
+	nextPri int64  // next priority level where more capacity will be available
+}
+
+// GetCapacityCurve returns a new or recently cached CapacityCurve based on the contents of the pool
+func (pp *PriorityPool) GetCapacityCurve() *CapacityCurve {
+	pp.lock.Lock()
+	defer pp.lock.Unlock()
+
+	now := pp.clock.Now()
+	dt := time.Duration(now - pp.ccUpdatedAt)
+	if !pp.ccUpdateForced && pp.cachedCurve != nil && dt < time.Second*10 {
+		return pp.cachedCurve
+	}
+
+	pp.ccUpdateForced = false
+	pp.ccUpdatedAt = now
+	curve := &CapacityCurve{
+		index: make(map[enode.ID][]int),
+	}
+	pp.cachedCurve = curve
+
+	var excludeID enode.ID
+	excludeFirst := pp.maxCount == pp.activeCount
+	// reduce node capacities or remove nodes until nothing is left in the queue;
+	// record the available capacity and the necessary priority after each step
+	for pp.activeCap > 0 {
+		cp := curvePoint{
+			freeCap: pp.maxCap - pp.activeCap,
```

Comment:
If the `pp.maxCap` is less than `pp.activeCap`, it will be very wrong. Although i think in theory it won't happen. 

Still, can we add this additional check? I'm not very confident to review it meaningfully, but let's not screw up the release at the last minute.

Reply:
It really shouldn't happen but now I print an error log if it does anyways :)